### PR TITLE
Changed python proto build to use shared protos

### PR DESCRIPTION
**Pull request by kpayson64**
Changed python proto build to use shared protos.  Also fixed up some proto syntax to make the build pass.  Right now the packaging solution for the generated protos is hacky, but based on #4123, it sounds like there was no clean solution.

### CODE REVIEW

File: src/proto/grpc/binary_log/v1alpha/log.proto [hunk 1]
```diff
@@ -29,20 +29,20 @@
 
 syntax = "proto3";
 
-import "google/protobuf/timestamp.proto"
```
Comment (jtattermusch) [inline src/proto/grpc/binary_log/v1alpha/log.proto]: @murgatroid99  can you please look at the changes in log.proto?

File: src/proto/grpc/testing/echo.proto [hunk 2]
```diff
@@ -32,7 +32,7 @@ syntax = "proto3";
 
 import "src/proto/grpc/testing/echo_messages.proto";
 
-package grpc.testing;
```
Comment (jtattermusch) [inline src/proto/grpc/testing/echo.proto]: why this change?

File: src/python/grpcio/commands.py [hunk 3]
```diff
@@ -157,31 +169,48 @@ def run(self):
...
       raise CommandError('Command:\n{}\nMessage:\n{}\nOutput:\n{}'.format(
           command, e.message, e.output))
 
+    for walk_root, _, _ in os.walk(proto_gen_stem):
+      path = os.path.join(walk_root, '__init__.py')
+      open(path, 'a').close()
+    with open(os.path.join(proto_gen_stem, '__init__.py'), 'w') as f:
+      f.write(PROTO_GEN_INIT)
+    
+    proto_stem =  os.path.join(GRPC_STEM, 'src', 'proto')
```
Comment (jtattermusch) [inline src/python/grpcio/commands.py]: did you forget to remove these two lines?

File: src/python/grpcio/commands.py [hunk 4]
```diff
@@ -157,31 +169,48 @@ def run(self):
...
     ] + paths
     try:
-      subprocess.check_output(' '.join(command), cwd=root_directory, shell=True,
+      subprocess.check_output(' '.join(command), cwd=PYTHON_STEM, shell=True,
                               stderr=subprocess.STDOUT)
     except subprocess.CalledProcessError as e:
       raise CommandError('Command:\n{}\nMessage:\n{}\nOutput:\n{}'.format(
           command, e.message, e.output))
 
+    for walk_root, _, _ in os.walk(proto_gen_stem):
```
Comment (jtattermusch) [inline src/python/grpcio/commands.py]: a hack like this definitely needs a comment that explains what the purpose is.

File: src/proto/grpc/testing/echo.proto (hunk 2 above)
Comment (kpayson64) [inline src/proto/grpc/testing/echo.proto]: There is a namspace collision on grpc.testing.UnimplementedService.  If you try to do a batch build of all the protos, it fails because of this.  I've reverted this and am doing singular builds because this namspace change broke the build.

File: src/python/grpcio/tests/interop/_insecure_interop_test.py [hunk 5]
```diff
@@ -31,12 +31,12 @@
 
 import unittest
 
+from gen.src.proto.grpc.testing import test_pb2
```
Comment (jtattermusch) [inline src/python/grpcio/tests/interop/_insecure_interop_test.py]: is there a way for this to stay `from grpc.testing import test_pb2`? I think that's actually a requirement.

File: src/proto/grpc/binary_log/v1alpha/log.proto (hunk 1 above)
Comment (murgatroid99) [inline src/proto/grpc/binary_log/v1alpha/log.proto]: These changes all look correct to me.

File: src/proto/grpc/binary_log/v1alpha/log.proto [hunk 6]
```diff
@@ -29,20 +29,20 @@
```
Comment (jtattermusch) [inline src/proto/grpc/binary_log/v1alpha/log.proto]: @kpayson64   if this PR gets delayed, can we at least create a separate PR to fix the log.proto? (it seems to be an independent change).

File: src/proto/grpc/binary_log/v1alpha/log.proto [hunk 7]
```diff
@@ -105,4 +105,4 @@ message Message {
   // The contents of the message. May be a prefix instead of the complete
   // message.
   bytes data = 5;
-}
```
Comment (nathanielmanistaatgoogle) [inline src/proto/grpc/binary_log/v1alpha/log.proto]: How certain are you that you intended to make the change on this line?

File: src/python/grpcio/commands.py [hunk 8]
```diff
@@ -157,30 +157,47 @@ def run(self):
     if not self.grpc_python_plugin_command:
       raise CommandError('could not find grpc_python_plugin '
                          '(protoc plugin for GRPC Python)')
+
+    proto_stem = os.path.join(GRPC_STEM, 'src', 'proto')
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/commands.py]: Does this value have to be computed in this function or can it be a module-scope constant?

File: src/python/grpcio/commands.py [hunk 9]
```diff
@@ -157,30 +157,47 @@ def run(self):
...
+    proto_gen_stem = os.path.join(GRPC_STEM, 'src', 'python', 'gens')
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/commands.py]: Can this be a constant?

File: src/python/grpcio/tests/interop/client.py [hunk 10]
```diff
@@ -28,15 +28,14 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 """The Python implementation of the GRPC interoperability test client."""
-
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/tests/interop/client.py]: Any reason not to retain the blank line between module doc string and imports?

File: tools/run_tests/run_tests.py [hunk 11]
```diff
@@ -371,6 +371,7 @@ def test_specs(self):
       tests_json = json.load(tests_json_file)
     environment = dict(_FORCE_ENVIRON_FOR_WRAPPERS)
     environment['PYVER'] = '2.7'
+    environment['PYTHONPATH'] = os.path.abspath('src/python/gens')
```
Comment (nathanielmanistaatgoogle) [inline tools/run_tests/run_tests.py]: This strikes me as a little brittle for the way that it sets the `PYTHONPATH` rather than simply appends to it. Should it look for an existing `PYTHONPATH` in the environment? Or am I being overly careful and fussy? Am I missing something that guarantees that there is no `PYTHONPATH` in the environment before getting to this line?

Comment (kpayson64) [inline tools/run_tests/run_tests.py]: When new tests are launched, they are launched with ```subprocess.Popen()``` with an explicit environment, which causes the child process to not inherit the current environment variables. 

Because test_specs() is more or less the entry point for a specific test, the only place we could have introduced a conflicting PYTHONPATH is with
```environment = dict(_FORCE_ENVIRON_FOR_WRAPPERS)```

File: src/python/grpcio/commands.py (hunk 8 above)
Comment (kpayson64) [inline src/python/grpcio/commands.py]: Changed

File: src/python/grpcio/commands.py (hunk 9 above)
Comment (kpayson64) [inline src/python/grpcio/commands.py]: Changed

File: src/proto/grpc/binary_log/v1alpha/log.proto (hunk 7 above)
Comment (kpayson64) [inline src/proto/grpc/binary_log/v1alpha/log.proto]: : |
I'm not even sure how to remove this.  The git cli won't register a diff like this, so even after 
```git checkout origin/master log.proto``` I'm willing to bet this still shows up in the github diff.

**Comment by kpayson64** [comment]
I've addressed all the comments

File: src/proto/grpc/binary_log/v1alpha/log.proto (hunk 7 above)
Comment (nathanielmanistaatgoogle) [inline src/proto/grpc/binary_log/v1alpha/log.proto]: I'm skeptical. If GitHub is showing us a diff, `git` must be smart enough to detect the diff.

Please remove the file from the pull request unless a "real" change needs to be made?

File: src/python/grpcio/commands.py [hunk 12]
```diff
@@ -59,7 +62,6 @@
 html_theme = 'sphinx_rtd_theme'
 """
 
-
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/commands.py]: Why drop this blank line?

File: src/python/grpcio/commands.py [hunk 13]
```diff
@@ -157,30 +159,45 @@ def run(self):
...
-        '--python-grpc_out={}'.format(root_directory),
-    ] + paths
-    try:
-      subprocess.check_output(' '.join(command), cwd=root_directory, shell=True,
-                              stderr=subprocess.STDOUT)
-    except subprocess.CalledProcessError as e:
-      raise CommandError('Command:\n{}\nMessage:\n{}\nOutput:\n{}'.format(
-          command, e.message, e.output))
+
+    # TODO(kpayson) it would be nice to do this in a batch command,
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/commands.py]: Most minor of minor points: missing colon after closing parenthesis on this line.

**Comment by kpayson64** [comment]
I've addressed all the comments.

**Comment by nathanielmanistaatgoogle** [comment]
I'm satisfied with the change content but I'd like @soltanmm to take a look as well.

Please squash commits.

File: src/python/.gitignore [hunk 14]
```diff
@@ -0,0 +1 @@
+gen/
```
Comment (soltanmm) [inline src/python/.gitignore]: Should this be `gen/` or `gens/`?

Comment (kpayson64) [inline src/python/.gitignore]: The C++ proto output directory is gens/, I figured I would be consistent with that.

Comment (soltanmm) [inline src/python/.gitignore]: I guess I was too vague with my nit. :-P I mean that in `commands.py` it appears that the folder is named `gens/` but here `gen/` is being ignored.

**Comment by soltanmm** [comment]
LGTM sans nit.

File: src/python/.gitignore (hunk 14 above)
Comment (kpayson64) [inline src/python/.gitignore]: Fixed

**Comment by soltanmm** [comment]
Python's green on the basic tests. Pipelining a ready-to-merge-label...

**Comment by jtattermusch** [comment]
mac node3 failure is unrelated.
